File: Dev/CS/Mascaret/Mascaret/VEHA/Behavior/StateMachine/StateMachineBehaviorExecution.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Mascaret
{
    public class StateMachineBehaviorExecution : BehaviorExecution
    {

        private Vertex currentState;
        public Vertex CurrentState
        {
            get { return currentState; }
            set { currentState = value; }
        }


        public bool currentActionDone;


        private bool toStop;
        public bool ToStop
        {
            get { return toStop; }
            set { toStop = value; }
        }


        private bool first;
        public bool First
        {
            get { return first; }
            set { first = value; }
        }


        private List<Trigger> eventQueue;
        public List<Trigger> EventQueue
        {
            get { return eventQueue; }
            set { eventQueue = value; }
        }


        private BehaviorExecution be;
        public BehaviorExecution Be
        {
            get { return be; }
            set { be = value; }
        }


        //default parameter sync=false
        public StateMachineBehaviorExecution(StateMachine stateMachine, InstanceSpecification host, Dictionary<string, ValueSpecification> p, bool sync)
            : base(stateMachine, host, p, sync)
        {
            first = true;
            currentActionDone = true;
            eventQueue = new List<Trigger>();
        }

        public void setCurrentState(string name)
        {
            Region region = getStateMachine().Region[0];
            foreach (Vertex currentVertex in region.Vertices)
            {
                if (currentVertex.name == name)
                    currentState = currentVertex;
            }
        }

        public bool simulateState(Vertex state)
        {
            currentState = state;
            state.simulate(Host);
            return true;
        }

        public StateMachine getStateMachine()
        {
            return (StateMachine)Specification;
        }

        public override double execute(double dt)
        {
          //  StreamWriter file = MascaretApplication.Instance.logfile; file.Flush();
         //   file.WriteLine("StateMachine"); file.Flush();
            if (first)
            {
                // TODO: Attention a gerer plus tard : possibilite de fixe l'etat en cours a l'initialisation de l'instance
                // Recherche de l'etat Initial
                List<PseudoState> ps = getStateMachine().ConnectionPoint;
                StateMachine machine = getStateMachine();
                List<Region> res = machine.Region;

                bool foundInitialState = false;
               // file.WriteLine("Pseudo state : " + ps.Count); file.Flush();
                foreach (PseudoState pseudo in ps)
                {
                    if (pseudo.kind == PseudoStateKind.INITIAL)
                    {
                        Dictionary<string, ValueSpecification> p = new Dictionary<string, ValueSpecification>();
                        List<Transition> trans = pseudo.Outgoing;
                        foreach (Transition transition in trans)
                        {
                            //if (_currentState) _currentState->desactivate();
                            activateState(transition.Target, p);
                        }
                        //file.WriteLine("Found Initial"); file.Flush();
                        foundInitialState = true;
                        break;
                    }
                }
                if (!foundInitialState) toStop = true;
                //toStop = foundInitialState ? false : true;
                first = false;
            }
            else
            {
                while (eventQueue.Count != 0)
                {
                    Vertex currentState = this.currentState;
                    if (currentState == null)//|| (currentState.Noninterpt && currentState.running)) 
                    { break; }

                   // if (this.Host.name == "tiroirProduits")
                       // file.WriteLine(" --->  " + this.Host.name + " stateMachine Behavior " + this.Specification.getFullName()+ " in " + currentState.getFullName()); file.Flush();

                    Trigger trigger = eventQueue[0];
                    eventQueue.RemoveAt(0);
                    List<Transition> trans = currentState.Outgoing;
                    MascaretApplication.Instance.VRComponentFactory.Log(" Trigger : " + trigger.name);
                   // file.WriteLine("Nb Trans : " + trans.Count); file.Flush();
                    for (int iT = 0; iT < trans.Count; iT++)
                    {
                        
                        List<Trigger> transTriggers = trans[iT].Trigger;
                      //  file.WriteLine("Nb Triger : " + transTriggers.Count); file.Flush();
                        for (int i = 0; i < transTriggers.Count; i++)
                        {
                            Trigger transTrigger = transTriggers[i];
                           // if (this.Host.name == "tiroirProduits")
                               // file.WriteLine(" --->  " + this.Host.name + " " +  transTrigger.getFullName()); file.Flush();
                            
                            if (transTrigger.equals(trigger))
                            {
                                Dictionary<string, ValueSpecification> param = new Dictionary<string, ValueSpecification>();
                                //OclParser::Context context;
                                MascaretApplication.Instance.VRComponentFactory.Log(" Trigger OK");

                                MascaretEvent evt = trigger.MEvent;
                                if (evt != null && evt.Type == "SignalEvent")
                                {
                                    SignalEvent sevt = (SignalEvent)evt;
                                    InstanceSpecification signal = sevt.Signal;
                                    if (signal != null)
                                    {
                                        /*  // Gestion des parametres
                                        map<string, shared_ptr<Slot> > props = signal->getSlots();
                                        map<string, shared_ptr<Slot> >::iterator it;
                                        for (it = props.begin(); it != props.end(); it++)
                                        {
                                            vector<AnyValue> values;
                                            values.push_back((AnyValue)it->second->getValue()->getStringFromValue());  
                                            context[it->first] = CollectionAnyValue(values);
                                            param[it->first] = it->second->getValue();
                                            cerr << it->first << " == " << it->second->getValue()->getStringFromValue() << endl;
                                        }
                                        */
                                    }
                                }
                                //context["self"]=getHost();

                                if (trans[iT].Guard == null) //|| trans[iT]->getGuard()->eval(context))
                                {
                                    if (currentState != null)
                                    {
                                        if (be != null)
                                        {
                                            be.stop();
                                            BehaviorScheduler.Instance.deleteExecutionBehavior(be);
                                            be = null;
                                        }
                                        currentState.desactivate();
                                    }

                                    Action effect = trans[iT].Effect;
                                    if (effect != null)
                                    {
                                        MascaretApplication.Instance.VRComponentFactory.Log("Has effect");

                                        BehaviorScheduler.Instance.executeBehavior(effect, this.Host, param, false);
                                    }

                                    MascaretApplication.Instance.VRComponentFactory.Log("Activate new state : " + trans[iT].Target.name);
                                    activateState(trans[iT].Target, param);
                                }
                            }
                        }
                    }
                }

                //	if(eventQueue.Count == 0)
                //		pause();
                //toStop = true;

            }

            return 0.01;
            //if (toStop) return 0;
            //else return 0.001;
        }

        public void addSignal(InstanceSpecification signal)
        {
           // StreamWriter file = MascaretApplication.Instance.logfile;
          //  file.WriteLine("recu signal : " + signal.name + " par " + Host.name); file.Flush();
            Trigger trigger = new Trigger(signal.name);
            SignalEvent se = new SignalEvent(signal.name);
            se.Signal = signal;
            trigger.MEvent = se;
            eventQueue.Add(trigger);
            //restart();
        }

        public void simulateAddSignal(InstanceSpecification signal)
        {

        }

        public void addTime(double time)
        {
            Trigger trigger = new Trigger("time");
            TimeEvent te = new TimeEvent("time");
            LiteralReal lValue = new LiteralReal(time);
            te.When = lValue;
            trigger.MEvent = te;
            eventQueue.Add(trigger);
            restart();
        }

        protected bool activateState(Vertex state, Dictionary<string, ValueSpecification> p)
        {
           // StreamWriter file = MascaretApplication.Instance.logfile;

          //  file.WriteLine("Activating State : " + state.name); file.Flush();
            currentState = state;

           if (state as FinalState == null)
                toStop = true;

            be = currentState.activate(Host, p);

            if (be != null)
            {

                currentState.running = true;
                //callBack function
                be.addCallbackOnBehaviorStop(onBehaviorStop);
                currentActionDone = false;
            }


            foreach (Transition currentTransition in currentState.Outgoing)
            {
                foreach (Trigger currentTrigger in currentTransition.Trigger)
                {
                    if (currentTrigger.MEvent.Type == "ChangeEvent")
                    {
                        ChangeEvent ce = currentTrigger.MEvent as ChangeEvent;
                        Expression exp = (Expression)ce.ChangeExpression.clone();
                        //callBack function
                        //evalutateExpression function
                    }
                }
            }


            return true;
        }

        protected void onBehaviorStop(BehaviorExecution be)
        {
            currentState.running = false;
            currentActionDone = true;

          //  StreamWriter file = MascaretApplication.Instance.logfile; file.Flush();
          //  file.WriteLine("Behavior stopped");


            foreach (Transition currentTransition in currentState.Outgoing)
            {
                if (currentTransition.Trigger.Count == 0)
                {
                    if (currentTransition.Guard == null) //|| trans[iT]->getGuard()->eval(context))
                    {
                        Dictionary<string, ValueSpecification> param = new Dictionary<string, ValueSpecification>();

                        Action effect = currentTransition.Effect;
                        if (effect != null)
                        {
                            BehaviorScheduler.Instance.executeBehavior(effect, this.Host, param, false);
                        }

                        activateState(currentTransition.Target, param);
                    }
                }
            }
        }

        protected void onExpressionCHange(Expression exp)
        {
            //expression and callBacks
        }
    }
}

File: Dev/C++/Mascaret/src/VEHA/Behavior/StateMachine/StateMachineBehaviorExecution.cpp
#include "VEHA/Behavior/Common/Signal.h"
#include "VEHA/Behavior/Common/Trigger.h"
#include "VEHA/Behavior/Common/behaviorScheduler.h"
#include "VEHA/Behavior/Common/behaviorScheduler.h"
#include "VEHA/Behavior/StateMachine/PseudoState.h"
#include "VEHA/Behavior/StateMachine/FinalState.h"
#include "VEHA/Behavior/StateMachine/SignalEvent.h"
#include "VEHA/Behavior/StateMachine/TimeEvent.h"
#include "VEHA/Behavior/StateMachine/ChangeEvent.h"
#include "VEHA/Behavior/StateMachine/Transition.h"
#include "VEHA/Behavior/StateMachine/StateMachine.h"
#include "VEHA/Behavior/StateMachine/Region.h"
#include "VEHA/Behavior/StateMachine/StateMachineBehaviorExecution.h"
#include "VEHA/Kernel/Constraint.h"
#include "VEHA/Kernel/InstanceSpecification.h"
#include "VEHA/Kernel/Expression.h"
#include <boost/bind.hpp>


using namespace boost;

namespace VEHA
{

class TimeCallbackBehavior : public BehaviorExecution
{
public :
	TimeCallbackBehavior(shared_ptr<StateMachine> stateMachine, shared_ptr<InstanceSpecification> host, const Parameters& p, double time,shared_ptr<BehaviorExecution> sme) :BehaviorExecution(stateMachine,host,p), _time(time), _sme(sme)
	{}
	void setTime(double time) {_time = time;}
	double execute(double dt) {(shared_dynamic_cast<StateMachineBehaviorExecution>(_sme))->addTime(_time); return 0;}
private :
	double _time;
	shared_ptr<BehaviorExecution> _sme;
};

StateMachineBehaviorExecution::StateMachineBehaviorExecution(
		shared_ptr<StateMachine> stateMachine, shared_ptr<InstanceSpecification> host, const Parameters& p, bool sync):BehaviorExecution(stateMachine,host,p,sync), _toStop (false)
{
	_first = true;
        currentActionDone = true;
}

StateMachineBehaviorExecution::~StateMachineBehaviorExecution()
{
}

void StateMachineBehaviorExecution::setCurrentState(string name)
{
   shared_ptr<Region> region = getStateMachine()->getRegion()[0];
   vector< shared_ptr<Vertex> > vertices = region->getVertices();
   for(unsigned int i = 0; i< vertices.size(); i++)
   {
      if (vertices[i]->getName() == name)
      {
	Parameters param;
        if (_be)
        {
          _be->stop();
          BehaviorScheduler::getInstance()->deleteExecutionBehavior(_be);
          _be=shared_ptr<BehaviorExecution>();
        }
        activateState(vertices[i], param);
      }
   }
}

void StateMachineBehaviorExecution::onExpressionChange(shared_ptr<Expression> exp)
{
	bool value=exp->evaluateExpression(getHost());
	if (value)
	{
		shared_ptr<Trigger> trigger( new Trigger("") );
		shared_ptr<ChangeEvent> ce( new ChangeEvent("") );
		ce->setChangeExpression(exp);
		trigger->setEvent(ce);
		_eventQueue.push_back(trigger);
		exp->removeCallbackOnMayChange(bind(&StateMachineBehaviorExecution::onExpressionChange,this,exp));
		BehaviorScheduler::getInstance()->reActivate(shared_from_this());
	}
}

void StateMachineBehaviorExecution::onBehaviorStop(BehaviorExecution * /*be*/)
{
	shared_ptr<Vertex> currentState=getCurrentState();
	vector<shared_ptr<Transition> > trans = currentState->getOutgoing();
	currentState->running = false;
        currentActionDone = true;
	for (size_t iT = 0; iT < trans.size(); iT ++)
	{
		vector<shared_ptr<Trigger> > transTriggers = trans[iT]->getTrigger();
		if (transTriggers.size() == 0)
		{
			OclParser::Context context;
                        context["self"]=getHost();
                        if(!trans[iT]->getGuard() || trans[iT]->getGuard()->eval(context))
                        {
                           // Launch transition effect
                           shared_ptr<Action> effect = trans[iT]->getEffect();
                           Parameters param;
                           if (effect)
                           {
                              BehaviorScheduler::getInstance()->executeBehavior(effect,getHost(),param);
                           }
			   activateState(trans[iT]->getTarget(), param);
                        }
		}
	}
}

bool StateMachineBehaviorExecution::activateState(shared_ptr<Vertex> state, const Parameters & param)
{
	_currentState = state;

	getHost()->fireChangeState(shared_dynamic_cast<StateMachineBehaviorExecution>(shared_from_this()));

        shared_ptr<FinalState> fs = shared_dynamic_cast<FinalState>(state);
        if (fs) _toStop = true;
	shared_ptr<Vertex> currentState=getCurrentState();
	_be = currentState->activate(getHost(), param);
	cerr << " Activate State : " << currentState->getName() << endl;
	
	if (_be)
	{
		currentState->running = true;
		_be->addCallbackOnBehaviorStop(bind(&StateMachineBehaviorExecution::onBehaviorStop,this,_1));
                currentActionDone = false;
	}

	vector<shared_ptr<Transition> > trans = currentState->getOutgoing();
	for (size_t iT = 0; iT < trans.size(); iT ++)
	{
		vector<shared_ptr<Trigger> > transTriggers = trans[iT]->getTrigger();
		for (size_t iTr = 0; iTr < transTriggers.size(); iTr ++)
		{
			shared_ptr<Event> event = transTriggers[iTr]->getEvent();
			if (event->getType() == "ChangeEvent")
			{
				shared_ptr<ChangeEvent> ce = shared_dynamic_cast<ChangeEvent>(event);
				shared_ptr<Expression> exp = shared_dynamic_cast<Expression>(ce->getChangeExpression()->clone());
				exp->addCallbackOnMayChange(bind(&StateMachineBehaviorExecution::onExpressionChange,this,exp));
				exp->evaluateExpression(getHost());
			}
/*
			else if (event->getType() == "TimeEvent")
			{
				shared_ptr<TimeEvent> te = shared_dynamic_cast<TimeEvent>(event);
				shared_ptr<LiteralReal> value = shared_dynamic_cast<LiteralReal>(te->getWhen());
				double timeValue = value->getValue();
				Parameters ps;
				shared_ptr<BehaviorExecution> be (new TimeCallbackBehavior(getStateMachine(),getHost(),ps,timeValue,shared_from_this()));
				BehaviorScheduler::getInstance()->_insert(be,timeValue);
			}
*/
		}
	}
	return true;
}

bool StateMachineBehaviorExecution::simulateState(shared_ptr<Vertex> state)
{
//cerr << "Simulate state : " << state->getName();
	_currentState = state;
	state->simulate(getHost());
	return true;
}

void StateMachineBehaviorExecution::addTime(double time)
{
	shared_ptr<Trigger> trigger( new Trigger("time") );
	shared_ptr<TimeEvent> te( new TimeEvent("time") );
	shared_ptr<LiteralReal> value (new LiteralReal(time));
	te->setWhen(value);
	trigger->setEvent(te);
	_eventQueue.push_back(trigger);
	restart();
}

void StateMachineBehaviorExecution::addSignal(shared_ptr<InstanceSpecification> signal)
{
	cerr << "recu signal : " << signal->getName() <<" par " << getHost()->getName()<< endl;
	shared_ptr<Trigger> trigger( new Trigger(signal->getName()) );
	shared_ptr<SignalEvent> se( new SignalEvent(signal->getName()) );
	se->setSignal(signal);
	trigger->setEvent(se);
	_eventQueue.push_back(trigger);
	restart();
}

void StateMachineBehaviorExecution::simulateAddSignal(shared_ptr<InstanceSpecification> signal)
{
//cerr << "Receive Signal : " << signal->getName() << endl;
	shared_ptr<Trigger> trigger( new Trigger(signal->getName()) );
	shared_ptr<SignalEvent> se( new SignalEvent(signal->getName()) );
	se->setSignal(signal);
	trigger->setEvent(se);

	shared_ptr<Vertex> currentState=getCurrentState();
	vector<shared_ptr<Transition> > trans = currentState->getOutgoing();
	for (size_t iT = 0; iT < trans.size(); iT ++)
	{
		vector<shared_ptr<Trigger> > transTriggers = trans[iT]->getTrigger();
		for (size_t i = 0; i < transTriggers.size() ; i++)
		{
			shared_ptr<Trigger> transTrigger = transTriggers[i];
			if (transTrigger->equals(trigger))
			{
				OclParser::Context context;
				context["self"]=getHost();
				if(!trans[iT]->getGuard() || trans[iT]->getGuard()->eval(context))
				{
					// Launch transition effect
					//shared_ptr<Action> effect = trans[iT]->getEffect();
					//if (effect)
					//{
				//		Parameters param;
				//		BehaviorScheduler::getInstance()->executeBehavior(effect,getHost(),param);
				//	}

					simulateState(trans[iT]->getTarget());
				}
			}
		}
	}
}

double StateMachineBehaviorExecution::execute(double dt)
{
	//cerr << " --->  " <<getHost()->getName()<<" stateMachine Behavior " << getSpecification()->getName() << endl;
	if (_first)
	{
		// TODO: Attention a gerer plus tard : possibilite de fixe l'etat en cours a l'initialisation de l'instance
		// Recherche de l'etat Initial
		vector<shared_ptr<PseudoState> > ps = getStateMachine()->getConnectionPoint();
		bool foundInitialState = false;
		for (size_t iPS = 0; iPS < ps.size(); iPS++)
		{
			if (ps[iPS]->getKind() == INITIAL)
			{
				Parameters p;
				vector<shared_ptr<Transition> > trans = ps[iPS]->getOutgoing();
				for (size_t iT = 0; iT < trans.size(); iT ++)
				{
					//if (_currentState) _currentState->desactivate();
					activateState(trans[iT]->getTarget(),p);
				}
				foundInitialState = true;
				break;
			}
		}
		_toStop = foundInitialState ? false : true;
		_first = false;
	}
	else
	{
		while (!_eventQueue.empty())
		{
			shared_ptr<Vertex> currentState=getCurrentState();
			if (!currentState || (currentState->getNonInterupt() && currentState->running)) {break;}

			shared_ptr<Trigger> trigger = _eventQueue.front();
			_eventQueue.pop_front();
			vector<shared_ptr<Transition> > trans = currentState->getOutgoing();
			for (size_t iT = 0; iT < trans.size(); iT ++)
			{
				vector<shared_ptr<Trigger> > transTriggers = trans[iT]->getTrigger();
				for (size_t i = 0; i < transTriggers.size() ; i++)
				{
					shared_ptr<Trigger> transTrigger = transTriggers[i];
					if (!transTrigger)
					{
						cerr << "[StateMachineBehaviorExecution Error] Find transition without Trigger in "
								<< getStateMachine()->getName() << endl;
					}
					if (transTrigger->equals(trigger))
					{
						Parameters param;
						OclParser::Context context;
                                                /* GetParams */
                                                shared_ptr<Event> evt = trigger->getEvent();
                                                if (evt && evt->getType() == "SignalEvent")
                                                {
                                                   shared_ptr<InstanceSpecification> signal = (shared_dynamic_cast<SignalEvent>(evt))->getSignal();
                                                   if (signal)
                                                   {
                                                      map<string, shared_ptr<Slot> > props = signal->getSlots();
                                                      map<string, shared_ptr<Slot> >::iterator it;
                                                      for (it = props.begin(); it != props.end(); it++)
                                                      {
                                                         vector<AnyValue> values;
                                                         values.push_back((AnyValue)it->second->getValue()->getStringFromValue());  
                                                         context[it->first] = CollectionAnyValue(values);
                                                         param[it->first] = it->second->getValue();
                                                         cerr << it->first << " == " << it->second->getValue()->getStringFromValue() << endl;
                                                      }
                                                   }
                                                }
						context["self"]=getHost();
      
						if(!trans[iT]->getGuard() || trans[iT]->getGuard()->eval(context))
						{
							if (currentState)
							{
								if (_be)
								{
									_be->stop();
									BehaviorScheduler::getInstance()->deleteExecutionBehavior(_be);
									_be=shared_ptr<BehaviorExecution>();
								} 
								currentState->desactivate();
							}
							// Launch transition effect
							shared_ptr<Action> effect = trans[iT]->getEffect();
							if (effect)
							{
								BehaviorScheduler::getInstance()->executeBehavior(effect,getHost(),param);
							}

							activateState(trans[iT]->getTarget(), param);
						}
					}
				}
			}
		}
		if(_eventQueue.empty())
			pause();
	}

        if (_toStop) return 0;
	else return 0.001;
}

} // Namespace
